### ISSUE TITLE
Support Union narrowing with Ellipsis

### ISSUE THREAD

**Issue by antdking** [microsoft/pyright#4397]
Effectively the same as [this issue from mypy](https://github.com/python/mypy/issues/13117)

Ellipsis is a [constant](https://docs.python.org/3/library/constants.html#Ellipsis) in Python, so we're able to do identity checks on it for narrowing.

This allows supporting using Ellipsis as a sentinel value where `None` is an accepted input.

basic example:

```py
import types

# crude, but valid in more complex cases, such as a fallback when accessing a store
def get_val(default=...) -> int | types.EllipsisType:
    return default

x = get_val()

if x is not ...:
    reveal_type(x)
```

reveal_type output:
```
Pylance: Type of "x" is "int | ellipsis"
mypy(note): Revealed type is "builtins.int"
```

**Comment by erictraut** [comment]
This doesn't strike me as a common idiom in Python, but since support for this type guard was added to mypy recently, I think it's reasonable to add support in pyright as well.

This will be addressed in the next release.

**Comment by antdking** [comment]
> Moving to pyright, since this is a core type checking issue.

Woops, sorry about that. Glad it's clean to move over.


> This doesn't strike me as a common idiom in Python

So you have another usecase (outside of "mypy did it"), I did more digging.
Our wider usecase is related to Pydantic, and how it represents [Required-Nullable fields](https://docs.pydantic.dev/usage/models/#required-optional-fields). Though I think you'll only encounter it when building dynamic models.

Thanks for the quick fix on this!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.288, which I just published. It will also be included in a future release of pylance.
